tutorials/haskell/golden-testing-in-haskell-with-hspec-gold/tutorial.md

```diff
@@ -0,0 +1,159 @@
+---
+title: Golden testing in Haskell with Hspec-Golden
+published: 2016-08-10
+tags: haskell, golden testing, snapshot testing, expect testing
+libraries: hspec-golden-1.0.0
+language: haskell
+author: Adrian Vera
+author-name: Adrian Vera
+twitter-profile: adrian_med
+github-profile: medAdrian
+description: A golden test is an IO action that writes its result to a file. To pass the test, this output file should be identical to the corresponding golden file, which contains the correct result for the test.
+---
+
+Golden tests are like unit tests, with the difference that the expected output is stored in a separate file. 
+
+Golden tests can also be known as:
+* Golden master testing
+* Approval testing
+* Snapshot testing
+* Expect testing
+
+There are many forms of getting the same result with different approaches, as the [Jest](https://jestjs.io/docs/en/snapshot-testing.html) snapshot testing,
+comparing two screenshots as [Wraith](https://github.com/BBC-News/wraith) or with the [Approval testing method](https://dzone.com/articles/approval-tests-alternative).
```

Comment:
@MedAdrian In these few paragraphs we're explaining a little bit what are golden tests. However, I think we should focus first on what is the problem that they try to solve, or the motivation behind them. I think we should answer the question: Do I have a problem that golden tests can help me with? I think that will help the reader to feel motivated about reading the rest of the tutorial. Let me know your thoughts.

Reply:
I think you are right, i've changed the conclusions section with the next paragraph:

``` 
We've learnt how to create golden tests thanks to hspec-golden and that this kind of testing will help us
verify modifications made to a reference version of our SUT(Subject Under Test) and validate that it has not modified its
behaviour in an incorrect way. Thus giving us some safety when extending and refactoring code that doesn't have proper tests.

When we write golden tests we build up knowledge of what the code is doing, this is very useful when we want to re-write it,
we can run our test to find out immediately whether we've modified the desired behaviour (we start to see our tests in a different way
since the purpose of golden testing is to document your system's behaviour, not to check for the wished one).

Once we adopt this perspective, we have a completely different frame for understanding automated tests in general,
we can see tests as descriptions of what we have rather than statements of correctness.
We can review our tests periodically to tighten up their conditions as we decide what the behavior should be at each level of our systems.
```

Reply:
@MedAdrian Looks good. Although, I think we should do some minor changes. Could you please add those paragraphs so that I can left some comments?

Reply:
Please refer to line 16 of the updated commit.